# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [2]:
os.getcwd()

'C:\\Users\\faggot\\Downloads\\1588204754_Step_2\\Step 2'

In [3]:
df = pd.read_csv('step3_output.csv')

df.head()

Unnamed: 0                 Name    state  summit_elev  vertical_drop  \
0           0       Alyeska Resort   Alaska         3939           2500   
1           1  Eaglecrest Ski Area   Alaska         2600           1540   
2           2     Hilltop Ski Area   Alaska         2090            294   
3           3     Arizona Snowbowl  Arizona        11500           2300   
4           4  Sunrise Park Resort  Arizona        11100           1800   

   base_elev  trams  fastEight  fastSixes  fastQuads  ...  SkiableTerrain_ac  \
0        250      1        0.0          0          2  ...             1610.0   
1       1200      0        0.0          0          0  ...              640.0   
2       1796      0        0.0          0          0  ...               30.0   
3       9200      0        0.0          1          0  ...              777.0   
4       9200      0        0.0          0          1  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         1  
1          53.0               90.0             0.0         1  
2          34.0              152.0            30.0         1  
3          89.0              122.0             0.0         2  
4          78.0              104.0            80.0         2  

[5 rows x 28 columns]

In [4]:
df = df.drop(columns=['Unnamed: 0'])

df.head()

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
0        0.0          0          2     2  ...             1610.0   
1        0.0          0          0     0  ...              640.0   
2        0.0          0          0     0  ...               30.0   
3        0.0          1          0     2  ...              777.0   
4        0.0          0          1     2  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         1  
1          53.0               90.0             0.0         1  
2          34.0              152.0            30.0         1  
3          89.0              122.0             0.0         2  
4          78.0              104.0            80.0         2  

[5 rows x 27 columns]

In [5]:
df.base_elev

0       250
1      1200
2      1796
3      9200
4      9200
       ... 
325    8500
326    6619
327    6237
328    8798
329    8400
Name: base_elev, Length: 330, dtype: int64

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [6]:
dfo=df['state'] 

df = pd.concat([df, pd.get_dummies(dfo)], axis=1)

In [7]:
df.head()

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  Rhode Island  South Dakota  \
0        0.0          0          2     2  ...             0             0   
1        0.0          0          0     0  ...             0             0   
2        0.0          0          0     0  ...             0             0   
3        0.0          1          0     2  ...             0             0   
4        0.0          0          1     2  ...             0             0   

   Tennessee  Utah  Vermont  Virginia  Washington  West Virginia  Wisconsin  \
0          0     0        0         0           0              0          0   
1          0     0        0         0           0              0          0   
2          0     0        0         0           0              0          0   
3          0     0        0         0           0              0          0   
4          0     0        0         0           0              0          0   

   Wyoming  
0        0  
1        0  
2        0  
3        0  
4        0  

[5 rows x 62 columns]

In [8]:
df = df.drop(columns=['state'])

df.head()

Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  Rhode Island  \
0        0.0          0          2     2       0  ...             0   
1        0.0          0          0     0       0  ...             0   
2        0.0          0          0     0       1  ...             0   
3        0.0          1          0     2       2  ...             0   
4        0.0          0          1     2       3  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 61 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [9]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [10]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [11]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [12]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test

y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [13]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [14]:
explained_variance_score(y_test, y_pred)

0.927753166174363

In [15]:
mean_absolute_error(y_test, y_pred)

5.210381803218575

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [16]:
lm.intercept_

64.13822583745056

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [17]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
df_model1 = pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

df_model1.sort_values(by=['Coefficient'],ascending=False)

Coefficient
total_chairs       1.728363e+13
fastQuads          6.552265e+12
surface            6.138981e+12
double             5.409898e+12
triple             4.826000e+12
quad               3.911295e+12
fastSixes          1.942422e+12
trams              1.668983e+12
fastEight          1.640774e+11
New York           1.055712e+11
Michigan           9.963056e+10
Colorado           8.778029e+10
California         8.590119e+10
Pennsylvania       8.197234e+10
Wisconsin          7.558492e+10
New Hampshire      7.558492e+10
Vermont            7.330122e+10
Minnesota          7.092803e+10
Utah               6.845602e+10
Montana            6.587407e+10
Idaho              6.587407e+10
Massachusetts      6.316871e+10
Oregon             6.032333e+10
Washington         6.032333e+10
New Mexico         5.731708e+10
Maine              5.731708e+10
Wyoming            5.412317e+10
North Carolina     4.701738e+10
Ohio               4.298699e+10
Connecticut        4.298699e+10
Nevada             3.850784e+10
Illinois           3.850784e+10
Virginia           3.850784e+10
West Virginia      3.850784e+10
Iowa               3.339987e+10
Alaska             3.339987e+10
Arizona            2.731255e+10
Indiana            2.731255e+10
South Dakota       2.731255e+10
Missouri           2.731255e+10
New Jersey         2.731255e+10
Rhode Island       1.934231e+10
Maryland           1.934231e+10
Tennessee          1.934231e+10
summit_elev        2.078652e+01
AdultWeekday       1.913843e+01
base_elev          1.905111e+01
vertical_drop      6.455674e+00
Runs               1.295349e+00
clusters           1.077637e+00
Snow Making_ac     9.237061e-01
NightSkiing_ac     7.587891e-01
LongestRun_mi      4.072266e-01
yearsOpen          3.825684e-01
averageSnowfall    2.460938e-01
TerrainParks       2.337189e-01
daysOpenLastYear   2.314453e-01
SkiableTerrain_ac  1.875000e-01
projectedDaysOpen  7.910156e-02

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [18]:
file="step3_output.csv"
df2=pd.read_csv(file)
X2=df2.drop(['AdultWeekend','Name','state'], axis=1)
y2=df2.AdultWeekend
scaler2 = preprocessing.StandardScaler().fit(X2)
X_scaled2=scaler2.transform(X2) 
y2=y2.ravel()
X_train2, X_test2, y_train2, y_test2 = train_test_split(X_scaled2, y2, test_size=0.25, random_state=1)
lm = linear_model.LinearRegression()
model2 = lm.fit(X_train2,y_train2)
y_pred2 = model2.predict(X_test2)
print(explained_variance_score(y_test2, y_pred2))
print(lm.intercept_)
mean_absolute_error(y_test2, y_pred2)

0.9149143155408435
64.08466748539195


5.68798838518573

In [19]:
df_model2 = pd.DataFrame(abs(lm.coef_), X2.columns, columns=['Coefficient'])  
df_model2.sort_values(by=['Coefficient'],ascending=False)

Coefficient
AdultWeekday         19.843831
summit_elev          13.087675
base_elev            10.338556
vertical_drop         4.347793
averageSnowfall       1.518968
triple                1.461242
quad                  1.352319
Runs                  1.219193
surface               1.077472
Unnamed: 0            1.046922
daysOpenLastYear      1.037671
clusters              0.872760
Snow Making_ac        0.794602
total_chairs          0.631571
fastQuads             0.566883
trams                 0.419084
double                0.337662
NightSkiing_ac        0.324431
yearsOpen             0.284830
TerrainParks          0.280517
SkiableTerrain_ac     0.253086
projectedDaysOpen     0.212463
fastSixes             0.190610
LongestRun_mi         0.168155
fastEight             0.138578

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [21]:
file="step3_output.csv"
df3=pd.read_csv(file)
X3=df3.drop(['Name','AdultWeekend','state','summit_elev','base_elev'], axis=1)
y3=df3.AdultWeekend
scaler3 = preprocessing.StandardScaler().fit(X3)
X_scaled3=scaler3.transform(X3) 
y3=y3.ravel()
X_train3, X_test3, y_train3, y_test3 = train_test_split(X_scaled3, y3, test_size=0.25, random_state=1)
lm = linear_model.LinearRegression()
model3 = lm.fit(X_train3,y_train3)
y_pred3 = model3.predict(X_test3)
print(explained_variance_score(y_test3, y_pred3))
print(lm.intercept_)
mean_absolute_error(y_test3, y_pred3)

0.9177400598121551
64.08511535742514


5.619939294005845

In [22]:
df_model3 = pd.DataFrame(abs(lm.coef_), X3.columns, columns=['Coefficient'])  
df_model3.sort_values(by=['Coefficient'],ascending=False)

Coefficient
AdultWeekday         19.795386
averageSnowfall       1.803409
triple                1.432393
quad                  1.345225
Runs                  1.298767
Unnamed: 0            1.182670
surface               1.015489
daysOpenLastYear      0.983964
vertical_drop         0.963566
clusters              0.755116
Snow Making_ac        0.741285
total_chairs          0.661079
fastQuads             0.618496
trams                 0.408313
NightSkiing_ac        0.401851
double                0.328717
yearsOpen             0.297945
SkiableTerrain_ac     0.295481
TerrainParks          0.209209
fastSixes             0.173257
LongestRun_mi         0.146759
projectedDaysOpen     0.022067
fastEight             0.001984

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.92775 | 5.2103 |-|
| Model 2. | 0.91491 | 5.6879 |'state'|
| Model 3. | 0.91774 | 5.6199 |'state','summit_elev','base_elev'|

Model Selection:

In [23]:
df_model3.sort_values(by=['Coefficient'],ascending=False)

Coefficient
AdultWeekday         19.795386
averageSnowfall       1.803409
triple                1.432393
quad                  1.345225
Runs                  1.298767
Unnamed: 0            1.182670
surface               1.015489
daysOpenLastYear      0.983964
vertical_drop         0.963566
clusters              0.755116
Snow Making_ac        0.741285
total_chairs          0.661079
fastQuads             0.618496
trams                 0.408313
NightSkiing_ac        0.401851
double                0.328717
yearsOpen             0.297945
SkiableTerrain_ac     0.295481
TerrainParks          0.209209
fastSixes             0.173257
LongestRun_mi         0.146759
projectedDaysOpen     0.022067
fastEight             0.001984